<a href="https://colab.research.google.com/github/Arup3201/deep-learning-materials/blob/main/1_Basics_of_Neural_Networks.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Understanding Neural Networks with Examples

### Neural Networks with a simple example

*Neural Network is a a type of system which is able to find out the relation between it's input and output.*

For instance:

* $X => -2, -1, 0, 1, 2, 3$
* $Y => -5, -3, -1, 1, 3, 5$

**Find the relation between X and Y. Try to derive the Y from it's corresponding X value. E.g. How can -5 be generated from -2?**

Take some time to think...


If you guessed the relation to be $Y=2X-1$ then congratulations!!

That's right!! You are now able to what neural network does the things...

Let's try to think how we can come up with this equation...

In [1]:
import numpy as np
import tensorflow as tf
from tensorflow import keras

In [ ]:
model = keras.Sequential([keras.layers.Dense(units=1, input_shape=[1])])

In [ ]:
model.compile(optimizer='sgd', loss='mean_squared_error')

In [ ]:
X = np.array([-2, -1, 0, 1, 2, 3], dtype=float)
Y = np.array([-5, -3, -1, 1, 3, 5], dtype=float)

# Y = 2X - 1

In [ ]:
model.fit(X, Y, epochs=100)

In [ ]:
model.predict([10.0])

1/1 [==============================] - 0s 49ms/step


array([[18.721066]], dtype=float32)

After learning from the examples, this model is able to predict the result for 10.0

Although the result is not exactly 19.0 but very close to 19.0

The reason is, our data is very small from which the model learnt so it was not able to learn everything and just from the 6 data points model will not have that much accuracy and confidence.

In [ ]:
model.get_weights()

[array([[1.9502501]], dtype=float32), array([-0.78143775], dtype=float32)]

From the result, you can see the coeffecient of X is very close to 2 and the bias value of close to -1.

### Another example to understand neural networks

In [ ]:
model = keras.Sequential([keras.layers.Dense(units=1, input_shape=[1])])
model.compile(optimizer='sgd', loss='mean_squared_error')

In [ ]:
xs = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], dtype=float)
ys = np.array([100, 150, 200, 250, 300, 350, 400, 450, 500, 550])

# Y = 50X+ 50

In [ ]:
model.fit(xs, ys, epochs=100)

In [ ]:
model.predict([15])

1/1 [==============================] - 0s 79ms/step


array([[831.8894]], dtype=float32)

In [ ]:
model.get_weights()

[array([[53.96724]], dtype=float32), array([22.380783], dtype=float32)]

### Neural Network with 2 inputs

In [ ]:
x1s = np.array([1, 2, 4, 6, 8, 10, 12], dtype=float)
x2s = np.array([2, 3, 4, 5, 6, 7, 8], dtype=float)
print(f"x1s shape={x1s.shape}")
print(f"x2s shape={x2s.shape}")

ys = np.array([ -6.5, -11.6, -14.7, -17.8, -20.9, -24. , -27.1], dtype=float)
print(f"ys shape={ys.shape}")

# Y = 2*X1 - 7.1*X2 + 5.7

x1s shape=(7,)
x2s shape=(7,)
ys shape=(7,)


In [ ]:
model = keras.Sequential([keras.layers.Dense(units=1, input_shape=[2])])
model.compile(optimizer='sgd', loss='mean_squared_error')

In [ ]:
X = np.concatenate([x1s[:, np.newaxis], x2s[:, np.newaxis]], axis=-1)

X.shape

(7, 2)

In [ ]:
X

array([[ 1.,  2.],
       [ 2.,  3.],
       [ 4.,  4.],
       [ 6.,  5.],
       [ 8.,  6.],
       [10.,  7.],
       [12.,  8.]])

In [ ]:
model.fit(X, ys, epochs=100)

In [ ]:
model.predict([[10, 11]])

1/1 [==============================] - 0s 97ms/step


array([[-36.595467]], dtype=float32)

In [ ]:
model.get_weights()

[array([[-0.1745129],
        [-3.0893433]], dtype=float32),
 array([-0.86755973], dtype=float32)]

The coefficient of X1 is -0.17 and X2 is -3.09 which shows that neural network is far from accurate and the reason is simple, in this case we had to come up with the equation containing 2 variables and given the dataset is small such task is even difficult for us.

### Neural Network for Classification Problem

In [6]:
xs = np.array([-10, -9, -8, -7, -6, -5, -4, -3,
               -2, -1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10], dtype=float)
ys = np.array([0, 0, 0, 0, 0, 0, 0, 0,
               0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1], dtype=float)

# 2*X - 5 > 0

In [8]:
model = keras.Sequential(keras.layers.Dense(1,
                                            activation='sigmoid',
                                            input_shape=[1]))

In [9]:
model.compile(optimizer='sgd', loss='binary_crossentropy')

In [ ]:
model.fit(xs, ys, epochs=100)

In [12]:
model.predict([10.0])

1/1 [==============================] - 0s 106ms/step


array([[0.9999854]], dtype=float32)

In [13]:
2*10-5 > 0

True

In [14]:
model.get_weights()

[array([[1.1250554]], dtype=float32), array([-0.11779122], dtype=float32)]